From: Hannes Reinecke <hare@suse.de>
Date: Mon, 15 Jun 2015 13:40:50 +0200
Subject: [PATCH] block: Discard bios do not have data
Git-commit: 458b76ed2f9517becb74dcc8eedd70d3068ea6e4 partial
Patch-Mainline: v3.14
References: bsc#928988

Discard bios have a payload, but this not counted as data.
So we need to update the 'bio_has_data()' check to correctly
account discard bios.
The corresponding upstream commit doesn't apply here as we're
missing the prerequisite patches.

Signed-off-by: Hannes Reinecke <hare@suse.de>
---
 include/linux/bio.h | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/include/linux/bio.h b/include/linux/bio.h
index d94b94a..31c2235 100644
--- a/include/linux/bio.h
+++ b/include/linux/bio.h
@@ -371,7 +371,8 @@ static inline char *__bio_kmap_irq(struct bio *bio, unsigned short idx,
  */
 static inline bool bio_has_data(struct bio *bio)
 {
-	if (bio && bio->bi_vcnt)
+	if (bio && bio->bi_vcnt &&
+	    (!bio->bi_rw & REQ_DISCARD))
 		return true;
 
 	return false;
-- 
1.8.5.2

